### ISSUE TITLE
Large prediction set when all classes but one have probability 0

### ISSUE THREAD

**Issue by christophM** [scikit-learn-contrib/MAPIE#254]
**Describe the bug**
Using MapieClassifier with method="cumulated_score" produces large prediction sets (that include all classes) for instances where all probabilities are 0 except for 1 class where it's 1 (e.g. [1,0,0] produces a set with all 3 classes).

**To Reproduce**

```Python
from mapie.classification import MapieClassifier
import numpy as np
from sklearn.datasets import load_iris
from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestClassifier

X,y = load_iris(return_X_y=True)
X_train, X_calib, y_train, y_calib = train_test_split(X, y, train_size=100, random_state=2)
model = RandomForestClassifier().fit(X_train, y_train)
mapie_score = MapieClassifier(model, cv="prefit", method="cumulated_score")
mapie_score.fit(X_calib, y_calib)
y_pred, y_set = mapie_score.predict(X_calib, alpha=0.05,include_last_label=False)
y_set = np.squeeze(y_set)
```

When I filter all the sets that have size==3, and look at the probability vector, I get this result:

```Python
all_classes = np.sum(y_set, axis=1) == 3
model.predict_proba(X_calib)[all_classes]
```
```
array([[0.03, 0.94, 0.03],
       [0.  , 1.  , 0.  ],
       [0.  , 0.  , 1.  ],
       [0.  , 1.  , 0.  ],
       [0.  , 1.  , 0.  ],
       [0.  , 0.  , 1.  ],
       [0.  , 1.  , 0.  ],
       [0.  , 1.  , 0.  ],
       [0.  , 0.  , 1.  ],
       [0.  , 1.  , 0.  ],
       [0.  , 0.  , 1.  ],
       [0.  , 1.  , 0.  ]])
```

**Expected behavior**
I would expect that these cases would result in prediction sets with just one class.

**Desktop (please complete the following information):**

- OS: MacOS
- Python (3.10.7)
- MAPIE Version 0.5.0

**Additional context**

This also occurred with the dry beans dataset, which has 7 classes and 13k instances. For the class that was simplest to classify, method="cumulated_score" would always produce sets containing all 7 classes. Both iris and dry beans have in common that they are "simple" in the sense that the model makes confident classifications with probability=1 for many instances.

Is that a bug, or intended behavior?
Seems like it has something to do with ties, but couldn't figure it out yet.

Thanks!

**Comment by christophM** [comment]
The same also happens when setting include_last_label=True.
When I pick include_last_label='randomized' one of the 0-probability classes is dropped.

**Comment by vincentblot28** [comment]
Dear Christoph, 

Thank you for your issue ! You are right about this problem. You have spotted a specific behavior that we didn't take into account. The reason of this error is the following:
When you use the cumulated score method, MAPIE sorts the outputs of the model in increasing order and do a cumsum over them. It then compares this cumsum to the quantile computed from the calibration conformity scores. When you have outputs with one value to 1 and all the others at 0, the cumsum is equal to one every where and always above the quantile. Hence, when we compute (line 410 in classification.py)
```python
y_pred_index_last = (
                    np.ma.masked_less(
                        y_pred_proba_cumsum
                        - threshold[np.newaxis, :],
                        -EPSILON
                    ).argmin(axis=1)
            )
```
the index which is returned is not necessarily the expected one and might correspond to a score of 0. Then when we do this computation (line 835 in classification.py)

```python
y_pred_proba_last = np.take_along_axis(
            y_pred_proba,
            y_pred_index_last,
            axis=1
        )
````
for observations with a 1 for a label and zeros everywhere else, it means that you will include all labels with a score greater than 0 in your prediction set. 

To resolve this issue (while waiting for the release), you can add this line bellow the above mentioned code :

```python
y_pred_proba_last[y_pred_proba_last <= EPSILON] = 1
```

We will keep you inform of the new release to correct this issue.

If you have any question about this or any other feature in MAPIE, do not hesitate :)

Vincent

**Comment by vincentblot28** [comment]
This PR should solve your issue : https://github.com/scikit-learn-contrib/MAPIE/pull/255

**Comment by christophM** [comment]
Wow, that was quick.

Thanks for the explanations and the fix.
